prow/cmd/deck/main.go

```diff
@@ -408,6 +435,202 @@ func handleBadge(ja *jobs.JobAgent) http.HandlerFunc {
+// A valid job GCS Job view url takes the form:
+//
+// /view/gcs/<bucketname>/<jobprefix>
+//
+// Example:
+// - /view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-large-performance/121
+func handleRequestGCSJobViews(sg *spyglass.Spyglass, ca *config.Agent) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		setHeadersNoCaching(w)
+		start := time.Now()
+		srcData := strings.TrimPrefix(r.URL.Path, "/view/gcs/")
+		src := fmt.Sprintf("gs://%s", srcData)
+
+		page, err := renderSpyglass(sg, ca, src)
+		if err != nil {
+			logrus.WithError(err).Error("error rendering spyglass page")
+			http.Error(w, "error getting views for job", http.StatusInternalServerError)
+			return
+		}
+
+		fmt.Fprint(w, page)
+		elapsed := time.Since(start)
+		logrus.WithField("duration", elapsed).Info("Loading view from GCS completed.")
+	}
+}
+
+// handleRequestJobViews handles requests to get all available artifact views for a given job via
+// a general src parameter, which can contain any string used to obtain job artifacts
+// A valid general job view url takes the form:
+//
+// /view/?src=<URI-encoded-source-string>
+//
+// Example:
+// - /view/?src=gs:%2F%2Fkubernetes-jenkins%2Fpr-logs%2Fpull%2Fkubeflow_kubeflow%2F1195%2Fkubeflow-presubmit%2F2558
+func handleRequestJobViews(sg *spyglass.Spyglass, ca *config.Agent) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		setHeadersNoCaching(w)
+		start := time.Now()
+		src := r.URL.Query().Get("src")
+		if src == "" {
+			http.Error(w, "missing src query parameter", http.StatusBadRequest)
+			return
+		}
+
+		page, err := renderSpyglass(sg, ca, src)
+		if err != nil {
+			logrus.WithError(err).Error("error rendering spyglass page")
+			http.Error(w, "error getting views for job", http.StatusInternalServerError)
+			return
+		}
+
+		fmt.Fprint(w, page)
+		elapsed := time.Since(start)
+		logrus.WithField("duration", elapsed).Info("Loading view from generic src completed.")
+	}
+}
+
+// renderSpyglass returns a pre-rendered Spyglass page from the given source string
+func renderSpyglass(sg *spyglass.Spyglass, ca *config.Agent, src string) (string, error) {
+	artifactNames, err := sg.ListArtifacts(src)
+	if err != nil {
+		return "", fmt.Errorf("error listing artifacts: %v", err)
+	}
+
+	viewerCache := map[string][]string{}
+	viewersRegistry := ca.Config().Deck.Spyglass.Viewers
+
+	for re, viewerNames := range viewersRegistry {
+		r, err := regexp.Compile(re)
+		if err != nil {
+			return "", fmt.Errorf("error compiling regexp %s: %v", r, err)
+		}
+		matches := []string{}
+		for _, a := range artifactNames {
+			if r.MatchString(a) {
+				matches = append(matches, a)
+			}
+		}
+		if len(matches) > 0 {
+			for _, vName := range viewerNames {
+				viewerCache[vName] = matches
+			}
+		}
+	}
+
+	viewsStart := time.Now()
+	lenses := sg.Views(viewerCache)
+	viewsElapsed := time.Since(viewsStart)
+	logrus.WithField("duration", viewsElapsed).Info("Loaded spyglass views")
```

Comment:
we're going to want identifying fields on these -- what was the request, what were the parameters, etc